### ISSUE TITLE
Adding roles to members using queue()

### ISSUE THREAD

**Issue by OmerBenGera** [DV8FromTheWorld/JDA#1211]
[download]: https://bintray.com/dv8fromtheworld/maven/JDA/_latestVersion
[guild]: https://discord.gg/0hMr4ce0tIk3pSjp
[stack overflow]: https://stackoverflow.com/questions/tagged/java

## General Troubleshooting

  
- [x] I have checked for similar issues.
- [x] I have updated to the [latest JDA version][download].
- [x] I have checked the branches or the maintainers' PRs for upcoming bug fixes.

## Bug Report

I tried to add methods to my guild members. It doesn't seem to work (or take a very very long time) to add the roles, when I am using the AuditableRestAction#queue method. I fixed it by using the AuditableRestAction#complete method, but that's not a good fix, as you know...

### Expected Behavior

It should add roles in a matter of seconds, and not waiting minutes for it to work.

### Code Example or Reproduction Steps

Not working:
`guild.addRoleToMember(member, role).queue()`
Working:
`guild.addRoleToMember(member, role).complete()`

### Exception or Error

N/A

**Comment by Andre601** [comment]
Please provide actual code snippets, so that we can see what could be the actual issue.

If you have more questions, don't hesitate to [join the JDA Discord](https://discord.gg/0hMr4ce0tIk3pSjp) and ask your questions there for a fast response.

**Comment by OmerBenGera** [comment]
> 
> 
> Please provide actual code snippets, so that we can see what could be the actual issue.
> 
> If you have more questions, don't hesitate to [join the JDA Discord](https://discord.gg/0hMr4ce0tIk3pSjp) and ask your questions there for a fast response.

What exactly do you need?
I am just adding the roles, and it doesn't work when using the queue method, but it works (same code) with the complete method.

**Comment by Andre601** [comment]
I just said what we need: The code you have.
Just saying "It doesn't work with X but with Y" doesn't help us at all to understand your setup.
We need to know what code you use to see if something there could cause this odd behaviour.

**Comment by MinnDevelopment** [comment]
Both of those work for me.

**Comment by MinnDevelopment** [comment]
I'm pretty sure the issue is a bug in your code and not JDA. The implementation for `complete()` just calls `queue()` and blocks the thread until the success callback is invoked.

**Comment by OmerBenGera** [comment]
> I'm pretty sure the issue is a bug in your code and not JDA. The implementation for `complete()` just calls `queue()` and blocks the thread until the success callback is invoked.

I will give more details and some code snippets later. I dont think its an issue on my end, as complete() works and with the same code, queue() doesnt.

**Comment by MinnDevelopment** [comment]
> I dont think its an issue on my end, as complete() works and with the same code, queue() doesnt.

Again, complete just calls queue and blocks the thread. So its not an issue with queue but an issue with your handling of asynchronous operations.

**Comment by bantu** [comment]
@OmerBenGera It appears this may have the same cause as #1212. Are you calling shutdown()?

**Comment by OmerBenGera** [comment]
First of all, no - I am not using the shutdown() method.
Here is the method that is being called:
`    public void addRolesToMember(Member member, Role... roles){
        long startTime = System.currentTimeMillis();
        if(member.getUser().getName().contains("Omer"))
            System.out.println("Received adding roles: " + Arrays.asList(roles));
        Arrays.stream(roles).forEach(role -> {
            guild.addRoleToMember(member, role).complete();
            if(member.getUser().getName().contains("Omer"))
                System.out.println("Took: " + (System.currentTimeMillis() - startTime));
        });
    }`
I added some debug messages so I can show you how much time it takes, here are some results:
with queue: https://gyazo.com/976e9f81cd095bbe6d034e12cab2a656
with complete: https://gyazo.com/1bd45eec6d02fcd9f1363150c0bc86d3

I am not calling the method above directly, I am calling:
`    public void updateUserRoles(VerifiedMember verifiedMember){
        List<Role> toAdd = new ArrayList<>();

        if(!verifiedMember.getProfiles().isEmpty())
            toAdd.add(verifiedRole);

        for(String pluginName : verifiedMember.getResources(false)){
            toAdd.add(pluginRoles.get(pluginName));
        }

        verifiedMember.addRoles(toAdd.toArray(new Role[]{}));
    }`
VerifiedMember#addRoles calls the method above.

**Comment by MinnDevelopment** [comment]
The code and logs you're providing make no sense to me. You're telling me the 5 lines here take 8 seconds with `complete()` and 360 seconds with `queue()`? This is ridiculous.

I'm still certain the real issue lies somewhere in your handling with asynchronous operations as per my previous comment. I recommend you use a real debugger and try to make a [mcve](https://stackoverflow.com/help/minimal-reproducible-example).

**Comment by OmerBenGera** [comment]
> 
> 
> The code and logs you're providing make no sense to me. You're telling me the 5 lines here take 8 seconds with `complete()` and 360 seconds with `queue()`? This is ridiculous.
> 
> I'm still certain the real issue lies somewhere in your handling with asynchronous operations as per my previous comment. I recommend you use a real debugger and try to make a [mcve](https://stackoverflow.com/help/minimal-reproducible-example).

I don't know if the issue is these actual methods, but you can clearly see the issue. Maybe the queue is full, idk - there are many tasks running at the same time, I really don't know.
I know complete() skips the queue, and queue() doesn't, which leads me to think it has to do with that. What do you think?

**Comment by MinnDevelopment** [comment]
> but you can clearly see the issue

Not really.

> I know complete() skips the queue

Nothing skips the queue. The raw `complete()` will simply append it to the queue and wait by blocking the thread.

If you do a lot of requests to the same endpoint, yes you will definitely be rate limited by discord. All rate limit backoff is logged on debug level in the `RateLimiter`. The endpoint for adding roles to a member has a 10/10s rate limit per guild.

**Comment by OmerBenGera** [comment]
> 
> 
> > but you can clearly see the issue
> 
> Not really.
> 
> > I know complete() skips the queue
> 
> Nothing skips the queue. The raw `complete()` will simply append it to the queue and wait by blocking the thread.
> 
> If you do a lot of requests to the same endpoint, yes you will definitely be rate limited by discord. All rate limit backoff is logged on debug level in the `RateLimiter`. The endpoint for adding roles to a member has a 10/10s rate limit per guild.

You can't see that it takes 360 seconds with queue and only 8 seconds with complete?
I showed code snippets and also debug messages.
What can I do then so you can understand the issue and the cause?

**Comment by Andre601** [comment]
> You can't see that it takes 360 seconds with queue and only 8 seconds with complete?
> I showed code snippets and also debug messages.
> What can I do then so you can understand the issue and the cause?

Do what Minn already suggested you to do:
> I recommend you use a real debugger and try to make a [mcve](https://stackoverflow.com/help/minimal-reproducible-example).

**Comment by MinnDevelopment** [comment]
> You can't see that it takes 360 seconds with queue and only 8 seconds with complete?

I don't believe queue will take 360 seconds to run. If that were the case, no bot written in JDA would be able to work. Additionally, I tested it myself and it took less than a second to run. I believe that the story here is incomplete, which is why I told you to use a real debugger (or profiler) and make a simple mcve.

**Comment by OmerBenGera** [comment]
> 
> 
> > You can't see that it takes 360 seconds with queue and only 8 seconds with complete?
> 
> I don't believe queue will take 360 seconds to run. If that were the case, no bot written in JDA would be able to work. Additionally, I tested it myself and it took less than a second to run. I believe that the story here is incomplete, which is why I told you to use a real debugger (or profiler) and make a simple mcve.

I made a new bot, with that code on startup:
```
            for(Role role : guild.getRoles()){
                try {
                    guild.addRoleToMember(member, role).queue(v -> {
                        System.out.println("Took " + (System.currentTimeMillis() - startTime));
                    });
                }catch(Exception ex){
                    continue;
                }

                counter++;

                if(counter == 10)
                    break;
            }
```
It doesn't seem to occur. Anything else I can do so you guys can better understand the issue?

**Comment by OmerBenGera** [comment]
> 
> 
> > You can't see that it takes 360 seconds with queue and only 8 seconds with complete?
> 
> I don't believe queue will take 360 seconds to run. If that were the case, no bot written in JDA would be able to work. Additionally, I tested it myself and it took less than a second to run. I believe that the story here is incomplete, which is why I told you to use a real debugger (or profiler) and make a simple mcve.

I can't use an actual debugger on a big bot, we won't be able to understand anything.
That's the only method so far that causes issues.
And if you believe it or not, it doesn't matter - that's a fact, it happens with the queue().

**Comment by MCOfficer** [comment]
> And if you believe it or not, it doesn't matter - that's a fact, it happens with the queue(), and complete() fixes iit.

Well... they are the people you are supposedly asking for a fix.

You can at least enable debug logs and see if you hit a lot of ratelimits or something similar.

**Comment by MinnDevelopment** [comment]
> It doesn't seem to occur. Anything else I can do so you guys can better understand the issue?

If it doesn't happen in a mcve then it must be some issue in your code. Try to isolate the issue by using a debugger. I'd say it's safe to say the issue is not with queue/complete but rather with something in your code, as I've alluded to since the beginning of this thread.

I recommend joining our discord server and figuring out the problem there. If you can find a real bug in JDA that is reproducible with a mcve then feel free to open a new issue.

[ ![discord-shield] ][discord-invite]

[discord-invite]: https://discord.gg/0hMr4ce0tIl3SLv5
[discord-shield]: https://discordapp.com/api/guilds/125227483518861312/widget.png

**Comment by Andre601** [comment]
> I can't use an actual debugger on a big bot, we won't be able to understand anything.
> That's the only method so far that causes issues.
> And if you believe it or not, it doesn't matter - that's a fact, it happens with the queue(), and complete() fixes iit.

It was mentioned before that `complete()` does essentially nothing more than `queue` the request and blocking the threat until the RestAction was performed.
We asked 3 times already for you to do a mcve and from what I can tell you didn't.
And your reason to not use a debugger on a big bot is kind of crap as most if not all IDEs have a debugger you can use.

**Comment by OmerBenGera** [comment]
I've also tested it with submit - and did it once with the queue parameter on true, and another time with false. When the parameter is true, it takes ages to respond. When the parameter is on false, it takes literally no time. It's your choice if you want to fix it or not, I found the solution :)

**Comment by OmerBenGera** [comment]
Working code, on the same bot that didn't work before:
```
    public void addRolesToMember(Member member, Role... roles){
        Arrays.stream(roles).forEach(role -> guild.addRoleToMember(member, role).submit(false));
    }
```

**Comment by DManstrator** [comment]
For the sake of completeness, there is a method which adds multiple roles to a member at once which is better than requesting an addition each. `Guild#modifyMemberRoles` is what you can use for that. Else I can just repeat what the others have said.

**Comment by MinnDevelopment** [comment]
> It's your choice if you want to fix it or not

I can't fix something that I can't reproduce and (probably) isn't actually broken.

**Comment by OmerBenGera** [comment]
> 
> 
> > It's your choice if you want to fix it or not
> 
> I can't fix something that I can't reproduce and (probably) isn't actually broken.

It was proven to be broken, as same code but with some other parameters work.

> 
> 
> For the sake of completeness, there is a method which adds multiple roles to a member at once which is better than requesting an addition each. `Guild#modifyMemberRoles` is what you can use for that. Else I can just repeat what the others have said.

Does this method add roles, or makes the roles of a member to be the ones that were provided?

**Comment by DManstrator** [comment]
> Does this method add roles, or makes the roles of a member to be the ones that were provided?
The documentation states that, look it up.

**Comment by MinnDevelopment** [comment]
> It was proven to be broken, as same code but with some other parameters work.

That doesn't prove anything, because the code is not isolated (which is what an mcve would give you). Asynchronous operations can be a bigger impact than synchronous operations on non-isolated code. If you want me to fix something you have to give me something to reproduce the problem. `submit(false)` and `complete(false)` will simply throw an exception if you are getting rate limited, causing it to never complete.
